CRM/MembershipExtras/Hook/Post/MembershipPayment.php

```diff
@@ -153,4 +176,150 @@ private function getRelatedRecurringLineItem() {
     return [];
   }
 
+  /**
+   * While Membership pre edit hook is responsible
+   * for creating periods upon membership renewal,
+   * the hook depends on the change on end date
+   * which get extended on renewal. But CiviCRM
+   * will not extend the end date of the membership
+   * if the membership  is renewed with pending contribution
+   * or payment plan and it will only be extended at the time
+   * of completing the contribution (or the first contribution
+   * in case of payment plan).
+   * But we need  an inactive period in this case and
+   * this method handles that.
+   */
+  private function createMissingMembershipPeriod() {
+    // membership payment post hook usually get called more than once for
+    // each single payment so we do this to prevent it from executing this
+    // method more than once for the same payment.
+    $counts = array_count_values(self::$paymentIds);
+    $hookCallCountsForId = $counts[$this->id];
+    if ($hookCallCountsForId > 1) {
+      return;
+    }
+
+    $contributionStatus = $this->contribution['contribution_status'];
+    if ($contributionStatus != 'Pending') {
+      return;
+    }
+
+    $isPaymentPlanPayment  = !empty($this->recurringContribution) && !empty($this->recurringContribution['installments']);
+    if($isPaymentPlanPayment) {

```

Comment:
the issue with this is that it will always call isFirstPaymentPlanContribution() and do the API
 call query no matter if it is a payment plan or not, I would prefer to avoid that, we have already many things that affect the performance and I would like to minimize that if it is possible.